src/core/net/netif.hpp

```diff
@@ -214,13 +214,15 @@ class Netif : public InstanceLocator, public LinkedListEntry<Netif>, private Non
     public:
         /**
          * This constructor initializes an `ExternalMulticastAddressIterator` instance to start from the first external
-         * multicast address.
+         * multicast address that matches a given Ip6 address type filter.
          *
-         * @param[in] aNetif  A reference to the Netif instance.
+         * @param[in] aNetif   A reference to the Netif instance.
+         * @param[in] aFilter  The Ip6 address type filter.
          *
          */
-        explicit ExternalMulticastAddressIterator(Netif &aNetif)
+        explicit ExternalMulticastAddressIterator(Netif &aNetif, Address::TypeFilter aFilter)
             : mNetif(aNetif)
+            , mFilter(aFilter)
```

Comment:
Not related to this PR, but an idea/suggestion maybe for a possible future enhancement:

We can consider declaring the iterator and iterator-builder classes as nested types within the `ExteranlNetifExternalMulticast` class. 
- This way we can use shorter names `Iterator` and `IteratorBuilder` for them in the implementation 
- But for users they are  properly name-spaced as `ExternalNetifMulticastAddress::Iterator`.

Reply:
I think it's a good idea. 
Even though usually `iterator` is an internal class of the container rather than entry (e.g. std::vector::iterator), I believe we don't have this option here.